### PR TITLE
Skill Calculator Expansion

### PR DESCRIPTION
**Pull request by TheStonedTurtle**
Thanks to @psikoi for helping to design the UI

## Skill Calculator Changes:
* Skill Planner Functionality added
* Banked Experience Functionality added
* Skill Calculator UI was updated to accommodate new functionality
* Added Config options to toggle new Functionality

## Planner Functionality:
* Right-Click an item to specify the amount you'd like to plan for
* Shows ending Level and Experience inside the **Target** box based on all planned actions.

## Banked Experience Functionality:
* Banked Experience shows exp from all items, regardless of required level
* Pulls inventory from entire bank regardless of visible tab
* Categorized items with toggle ability inside panel
* Shows ending Level and Experience inside the **Target** box
* Currently only supports single item training methods and fixed xp rates per raw resource
* Banked Experience currently works for:
	* Construction
	* Cooking
	* Crafting
	* Prayer
	* Smithing

## API Updates:
* Updated the **Skill** api to have a **getByName** method which returns the Skill if the passed string matches the name. All strings are upper-cased for compatibility.

## UI Component Updates:
* Added components to assist in creating **JComboBox** options with an icon and text.
* Updated **MaterialTab** to have RuneScape font as the default and to allow setting of Font, Selected Border, and Unselected Border.

## Summary of new UI Components
<details>
<p>

### ComboBoxIconEntry
* Requires an **Icon** and **String**.
* Should be added to your **JComboBox**
### ComboBoxIconRender
* Used for rendering **ComboBoxIconEntry** elements on a **JComboBox**
* Should be passed to **JComboBox.setRenderer()** 
* Supports no default selection (Must select the -1 index or null item, defaults to selecting first option)
* Settable default text for no default selection. (Default Value: "Select an option...")

</p>
</details>

## Media
<details>
<p>

#### Banked Experience

</p>
</details>

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/ui/components/materialtabs/MaterialTab.java [hunk 1]
```diff
@@ -49,11 +51,11 @@
  */
 public class MaterialTab extends JLabel
 {
-	private static final Border SELECTED_BORDER = new CompoundBorder(
+	private Border SELECTED_BORDER = new CompoundBorder(
```
Comment (psikoi) [inline runelite-client/src/main/java/net/runelite/client/ui/components/materialtabs/MaterialTab.java]: If these are no longer constants, use camel case.

File: runelite-client/src/main/java/net/runelite/client/ui/components/materialtabs/MaterialTab.java [hunk 2]
```diff
@@ -67,12 +69,15 @@
 	@Setter
 	private Runnable onSelectEvent;
 
+	private Font FONT = FontManager.getRunescapeFont();
```
Comment (psikoi) [inline runelite-client/src/main/java/net/runelite/client/ui/components/materialtabs/MaterialTab.java]: use camel case for non-constant variables

File: runelite-client/src/main/java/net/runelite/client/ui/components/materialtabs/MaterialTab.java [hunk 3]
```diff
@@ -162,6 +168,27 @@ public void unselect()
...
+	}
+
+	public void setUnselectedBorder(Border border)
+	{
+		UNSELECTED_BORDER = border;
+		if (!selected)
+			setBorder(UNSELECTED_BORDER);
+	}
+
+	public void setDefaultFont(Font font)
```
Comment (psikoi) [inline runelite-client/src/main/java/net/runelite/client/ui/components/materialtabs/MaterialTab.java]: you can just override setFont

**Comment by psikoi** [comment]
Don't make new commits for those things, edit previous commits instead

**Comment by psikoi** [comment]
:( rip

**Comment by FrontEndLou** [comment]
pretty new to github and development in general so excuse my ignorance, but what ever happened to this? I had the same idea and actually planned on using this idea as a starter project for myself

**Comment by TheStonedTurtle** [comment]
I believe the drop down was decided against, the planner didn't work for magic, and there are websites that do most of it better. Banked exp is stupid complex and doesn't look good in such a small UI, there are websites you can use with a little work.

**Pull request closed without merging**